src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,250 @@
+      x_block_padded(2*i+1) = tmp * h(2*i+1) + x_block_padded(2*i+1) * h(2*i);
+    }
+
+    srfft_->Compute(x_block_padded.Data(), false);
+    x_block_padded.Scale(1.0/N);
+
+    // combine the block
+    if (po + L < xdim) {       // current block is not the last block
+      for (int j = 0; j < L; j++)
+        x(po + j) = x_block_padded(j);
+      for (int j = 0; j < M-1; j++)
+        x(po + j) += temp_pad(j);
+      for (int j = 0; j < M-1; j++)
+        temp_pad(j) = x_block_padded(j + L);
+    } else {
+      for (int j = 0; j < xdim - po; j++)
+        x(po + j) = x_block_padded(j);
+      for (int j = 0; j < M-1 && j < xdim - po; j++)
+        x(po + j) += temp_pad(j);
+    }
+  }
+}
+
+void RepeatedlyAdd(Vector<BaseFloat> &x, Vector<BaseFloat> &h) {
+  for (int po = 0; po < x.Dim(); po += h.Dim()) {
+    int remain = x.Dim() - po;
+    int l = h.Dim();
+    if ( remain < l) l = remain;
+    SubVector<float> x_block(x, po, l);
+    SubVector<float> subh(h, 0, l);
+    x_block.AddVec(1.0, subh);
+  }
+}
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+
+    const char *usage =
+        "Corrupts the wave files supplied via input pipe with the specified\n"
+        "room-impulse response (RIR) and additive noise distortions\n"
+        "(specified by corresponding files).\n";
+
+    ParseOptions po(usage);
+    std::string noise_file;
+    BaseFloat snr_db = 20;
+
+    po.Register("noise-file", &noise_file,
+                "file with additive noise");
+    po.Register("snr-db", &snr_db,
+                "desired SNR(dB) of the output");
+
+    po.Read(argc, argv);
+    if (po.NumArgs() != 3) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string source_file = po.GetArg(1);
+    std::string rir_file = po.GetArg(2);
+    std::string out_file = po.GetArg(3);
+
+    WaveData src_wave;
+    {
+      Input ki(source_file);
+      src_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &SRC = src_wave.Data();
+    BaseFloat samp_freq_src = src_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_src = SRC.NumCols(),  // number of samples in recording
+          num_chan_src = SRC.NumRows();  // number of channels in recording
+
+    KALDI_LOG << "samp_freq of source is " << samp_freq_src
+              << " num_samp " << num_samp_src
+              << " num_chan " << num_chan_src;
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &RIR = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_rir = RIR.NumCols(),  // number of samples in recording
+          num_chan_rir = RIR.NumRows();  // number of channels in recording
+    KALDI_LOG << "samp_freq of rir is " << samp_freq_rir
+               << " num_samp " << num_samp_rir
+               << " num_chan " << num_chan_rir;
+
+    Vector<BaseFloat> src(num_samp_src);
```

Comment:
>Just to clarify,
>when --multi-channel-output is true, the output wave will contain multiple channels and the number of channels equals the number of channels of the RIR file, correct?

Yes

>By the way, if we allow users to select different channels in the RIR and noise file, can we skip verifying the RIR and noise file to have same number of channels? This provides users for having a greater flexibility in the combination of RIR and noise.

Channel selection should only supported in single channel output mode. If they don't have the same number of channels the RIR and noise are probably not recorded in the same setup. In this case I don't know if it even makes sense to add the noise after reverberation. However lets leave this choice to the user for now. Do not check that the RIR and noise files have same number of channels in single channel mode.